src/shared/utilities/disposableFiles.ts

```diff
@@ -0,0 +1,100 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as del from 'del'
+import * as fs from 'fs'
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import * as filesystem from '../filesystem'
+
+export class DisposableFiles implements vscode.Disposable {
+
+    private _disposed: boolean = false
+    private readonly _filePaths: Set<string> = new Set<string>()
+    private readonly _folderPaths: Set<string> = new Set<string>()
+
+    public addFile(file: string): DisposableFiles {
+        this._filePaths.add(file)
+
+        return this
+    }
+
+    public addFolder(file: string): DisposableFiles {
+        this._folderPaths.add(file)
+
+        return this
+    }
+
+    public dispose(): void {
+        if (!this._disposed) {
+            del.sync(
+                [...this._filePaths],
+                {
+                    absolute: true,
+                    force: true,
+                    nobrace: false,
+                    nodir: true,
+                    noext: true,
+                    noglobstar: true,
+                })
+
+            this._folderPaths.forEach(folder => {
+                if (fs.existsSync(folder)) {
+                    del.sync(
+                        folder,
+                        {
+                            absolute: true,
+                            force: true,
+                            nobrace: false,
+                            nodir: false,
+                            noext: true,
+                            noglobstar: true,
+                        })
+                }
+            })
+
+            this._disposed = true
+        }
+    }
+}
+
+export class ExtensionDisposableFiles {
+    protected static INSTANCE: DisposableFiles = new DisposableFiles()
```

Comment:
`readonly`

Reply:
It is protected, and not readonly so that unit tests can 'reset' it